text/0000-url-primitives.md

```diff
@@ -0,0 +1,349 @@
+
+    return `/${url}?${query}`; // => /blogs/1/posts/2?foo=bar
+  }
+}
+```
+
+### i18n routes
+In this scenario, we may want to internationalized route segments.
+
+For example, in English, we may want a route to be `/blogs/1/posts/2`, 
+but in Korean, `/블로그/1/게시물/2`.
+
+```ts
+// app/router.js/ts
+class CustomURLManager extends URLManager {
+  // current language: Korean
+  @service i18n;
+  @service router;
+
+  fromURL(url: string): RouteInfo { // /블로그/1/게시물/2?foo=bar
+    let [path, query] = url.split('?');
+    let segments = path.split('/');
+
+    // this dosen't have to be english, but it does have to match the names as
+    // defined in Router.map(...);
+    let english = 
+      path.split('/')
+          .segments.map(segment => {
+            return this.i18n.lookup(`routes.from.${segment}`, 'en-us') || segment;
+          })
+          .join('/');
+
+    let routeInfo = this.router.recognize(english);
+    let queryParams = qs.parse(query);
+
+    return {
+      ...routeInfo,
+      queryParams,
+    };
+  }
+
+  toURL(routeInfo: RouteInfo) {
+    let { 
+      mapInfo: {
+        segments // [blogs, :blogId, posts, :postId]
+      },
+      queryParams, // { foo: 'bar' }
+      dynamicSegments,
+    } = routeInfo;
+
+    let url = segments.map(segment => {
+      return dynamicSegments[segment] || this.i18n.t(`routes.${segment}`);
+    }).join('/');
+
+    let query = qs.stringify(queryParams);
+
+    return `/${url}?${query}`; // => /블로그/1/게시물/2?foo=bar
+  }
+}
+
+
+```
+
+### Custom URL Management per-route
+There are situations in which links and their routes don't conform to the rest of the application. For these situations, here is an example showing how manage the URL separately for those routes.
+
+```ts
+// app/router.js/ts
+Router.map(function() {
+  this.route('no-query-params', {
+    fromURL(url: string): RouteInfo {
+      let [path, query] = url.split('?');
+      
+      return this.router.recognize(path);
+    },
+    toURL(routeInfo: RouteInfo, dynamicSegments: object) {
+      return this.router.urlFor({routeInfo, queryParams: undefined });
+    }
+  });
+  // this.route('posts', ...)
+});
+
+class CustomURLManager extends URLManager {
+  @service router;
+
+  fromURL(url: string): RouteInfo {
+    let [path, query] = url.split('?');
+    // NOTE: this method could be named anything, as long as it matches
+    //       what is in the options / mapInfo object of the Router.map
+    let { fromURL } = this.router.mapInfoFrom(path)
+
+    if (fromURL) return fromURL(url);
+
+    // ...
+  }
+
+  toURL(routeInfo: RouteInfo, dynamicSegments: object) {
+    let { mapInfo: { toURL } } = routeInfo;
+
+    if (toURL) return toURL(routeInfo, dynamicSegments);
+
+    // ...
+  }
+}
+```
+
+## Detailed design
+
+### Additions to existing Public APIs
+1. add full list of dynamicSegments to RouteInfo so that the task of building out the map of dynamic segments to their values isn't in user-space
```

Comment:
```suggestion

1. Add full list of dynamicSegments to `RouteInfo` so that the task of building out the map of dynamic segments to their values isn't in userspace;
```